services/placement/algo/helper.go

```diff
@@ -296,32 +317,39 @@ func (ph placementHelper) getShardLen() int {
 }
 
 func (ph placementHelper) canAssignInstance(shard uint32, from, to services.PlacementInstance) bool {
-	if to.Shards().ContainsShard(shard) {
+	if to.Shards().Contains(shard) {
 		return false
 	}
 	return ph.opts.LooseRackCheck() || ph.HasNoRackConflict(shard, from, to.Rack())
 }
 
-func (ph placementHelper) assignShardToInstance(shard uint32, to services.PlacementInstance) {
+func (ph placementHelper) assignShardToInstance(s shard.Shard, to services.PlacementInstance) {
 	if to == nil {
 		return
 	}
 
-	to.Shards().AddShard(shard)
+	to.Shards().Add(s)
 
-	if _, exist := ph.shardToInstanceMap[shard]; !exist {
-		ph.shardToInstanceMap[shard] = make(map[services.PlacementInstance]struct{})
+	if _, exist := ph.shardToInstanceMap[s.ID()]; !exist {
+		ph.shardToInstanceMap[s.ID()] = make(map[services.PlacementInstance]struct{})
 	}
-	ph.shardToInstanceMap[shard][to] = struct{}{}
+	ph.shardToInstanceMap[s.ID()][to] = struct{}{}
 }
 
-func (ph placementHelper) removeShardFromInstance(shard uint32, from services.PlacementInstance) {
+func (ph placementHelper) removeShardFromInstance(s shard.Shard, from services.PlacementInstance) {
 	if from == nil {
 		return
 	}
 
-	from.Shards().RemoveShard(shard)
-	delete(ph.shardToInstanceMap[shard], from)
+	if s.State() == shard.Initializing {
+		from.Shards().Remove(s.ID())
+	} else if s.State() == shard.Available {
+		s.SetState(shard.Leaving)
+	} else if s.State() == shard.Leaving {
+		panic("should never move a leaving shard")
```

Comment:
Some of this code will likely one day run inside a discovery service no?  If so I think the correct thing here is to return an error and abandon a change rather than panic and bring down a running service.